File: Aran/Panda/EnRoute/User Interface/AranPlugin.vb
Imports Microsoft.Win32
Imports System.Collections.Generic
Imports Aran
Imports Aran.Aim
Imports Aran.AranEnvironment

<System.Runtime.InteropServices.ComVisibleAttribute(False)> Public Class ArrivalAranPlugin
	Inherits AranPlugin

	Const _name As String = "PANDA  -  EnRoute"

	Dim _tsmiEnRoute As ToolStripMenuItem

#Region "IAranPlugin Members"

	Public Sub New()
		'_isInit = False
	End Sub

	Public Overrides ReadOnly Property Id As Guid
		Get
			Return New Guid("160D6C2F-5547-4A40-93A1-C9A0198565AE")
		End Get
	End Property

	Public Overrides ReadOnly Property Name As String
		Get
			Return _name
		End Get
	End Property

	Public Overrides Function GetLayerFeatureTypes() As List(Of FeatureType)
		Dim featTypes As List(Of FeatureType) = New List(Of FeatureType)()

		featTypes.Add(FeatureType.AirportHeliport)
		featTypes.Add(FeatureType.VerticalStructure)
		featTypes.Add(FeatureType.VOR)
		featTypes.Add(FeatureType.DME)
		featTypes.Add(FeatureType.NDB)
		featTypes.Add(FeatureType.InstrumentApproachProcedure)
		featTypes.Add(FeatureType.DesignatedPoint)
		featTypes.Add(FeatureType.RunwayCentrelinePoint)

		Return featTypes
	End Function

	Public Overrides Sub Startup(aranEnv As AranEnvironment.IAranEnvironment)
		gAranEnv = aranEnv
		gHookHelper = New ESRI.ArcGIS.Controls.HookHelper
		gHookHelper.Hook = gAranEnv.HookObject

        LangCode = 1033
		SetThreadLocale(LangCode)
		My.Resources.Culture = New System.Globalization.CultureInfo(LangCode)


		Dim menuItem As New ToolStripMenuItem()
        menuItem.Text = "PANDA  -  Enroute" 'My.Resources.str80
		_tsmiEnRoute = New ToolStripMenuItem()
		_tsmiEnRoute.Text = My.Resources.str0002	'My.Resources.str158	'PrecApproach
		_tsmiEnRoute.Tag = 1
		AddHandler _tsmiEnRoute.Click, AddressOf Enroute_Click
		menuItem.DropDownItems.Add(_tsmiEnRoute)


		aranEnv.AranUI.AddMenuItem(AranMapMenu.Applications, menuItem)
	End Sub
#End Region

	Public Function InitCommands() As Boolean
		Return InitCommand() > 0
	End Function

	Public Sub Enroute_Click(sender As Object, e As EventArgs)


		Try
			Aran.PANDA.Common.NativeMethods.ShowPandaBox(GlobalVars.gAranEnv.Win32Window.Handle.ToInt32())

			If InitCommands() Then
				Dim EnrouteFrm As CEnrouteForm = New CEnrouteForm

				AddHandler EnrouteFrm.FormClosed, AddressOf AllForms_FormClosed

				_tsmiEnRoute.Checked = True
				EnrouteFrm.Show(s_Win32Window)
				'EnrouteFrm.ComboBox0001.Focus()
				EnrouteFrm = Nothing
			End If

			Aran.PANDA.Common.NativeMethods.HidePandaBox()
		Catch ex As Exception
			Aran.PANDA.Common.NativeMethods.HidePandaBox()
			Throw ex
		End Try
	End Sub

	Private Sub AllForms_FormClosed(sender As Object, e As FormClosedEventArgs)
		_tsmiEnRoute.Checked = False
		_tsmiEnRoute.Enabled = True

		sender = Nothing
		GC.Collect()
	End Sub
End Class

File: Aran/Panda/Departure/trunk/AranPlugin.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Aran.AranEnvironment;
using Aran.PANDA.Departure.Properties;
using Microsoft.Win32;
using Aran.Aim;
using Aran.Aim.Data;

namespace Aran.PANDA.Departure
{
	public class DeparaturePlugin : AranPlugin
	{
		private const string _name = "PANDA Conventional - Departure";
		//private bool _isInitCommands;
		private ToolStripMenuItem _tsmiDepartOmniDirect;
		private ToolStripMenuItem _tsmiDepartRouts;
		private ToolStripMenuItem _tsmiDepartGuidance;

		#region IAranPlugin Members

		public DeparaturePlugin()
		{
			//_isInitCommands = false;
		}

		public override Guid Id { get { return new Guid("c3860b1d-89c8-4fc5-9477-c105d223e775"); } }
		public override string Name { get { return _name; } }

		public override List<FeatureType> GetLayerFeatureTypes()
		{
			var list = new List<FeatureType>();

			list.Add(FeatureType.AirportHeliport);
			list.Add(FeatureType.RunwayCentrelinePoint);
			list.Add(FeatureType.DesignatedPoint);
			list.Add(FeatureType.VOR);
			list.Add(FeatureType.NDB);
			list.Add(FeatureType.DME);
			list.Add(FeatureType.VerticalStructure);

			return list;
		}

		public override void Startup(IAranEnvironment aranEnv)
		{
			GlobalVars.gAranEnv = aranEnv;
			GlobalVars.gHookHelper = new ESRI.ArcGIS.Controls.HookHelper();
			GlobalVars.gHookHelper.Hook = GlobalVars.gAranEnv.HookObject;

			int LangCode = 1033;
			NativeMethods.SetThreadLocale(LangCode);
			Resources.Culture = new System.Globalization.CultureInfo(LangCode);

			ToolStripMenuItem menuItem = new ToolStripMenuItem();
			menuItem.Text = Resources.str00080;

			_tsmiDepartOmniDirect = new ToolStripMenuItem();
			_tsmiDepartOmniDirect.Text = Resources.str15063;
			_tsmiDepartOmniDirect.Tag = 0;
			_tsmiDepartOmniDirect.Click += DepartOmniDirect_Click;
			menuItem.DropDownItems.Add(_tsmiDepartOmniDirect);

			_tsmiDepartRouts = new ToolStripMenuItem();
			_tsmiDepartRouts.Text = Resources.str15270;
			_tsmiDepartOmniDirect.Tag = 1;
			_tsmiDepartRouts.Click += DepartRouts_Click;
			menuItem.DropDownItems.Add(_tsmiDepartRouts);

			_tsmiDepartGuidance = new ToolStripMenuItem();
			_tsmiDepartGuidance.Text = Resources.str15478;
			_tsmiDepartOmniDirect.Tag = 2;
			_tsmiDepartGuidance.Click += DepartGuidance_Click;
			menuItem.DropDownItems.Add(_tsmiDepartGuidance);

			aranEnv.AranUI.AddMenuItem(AranMapMenu.Applications, menuItem);
		}

		#endregion

		private void InitCommands()
		{
			//if (_isInitCommands)
			//   return;

			GlobalVars.InitCommand();

			//_isInitCommands = true;
		}

		#region Event handlers
		private void DepartOmniDirect_Click(object sender, EventArgs e)
		{
			if (GlobalVars.CurrCmd == 1)
				return;
			GlobalVars.CurrCmd = 0;

			try
			{
                if (!SlotSelector())
                    return;

				NativeMethods.ShowPandaBox(GlobalVars.gAranEnv.Win32Window.Handle.ToInt32());

				InitCommands();

				CDepartOmniDirect DepartOmniDirectFrm = new CDepartOmniDirect();

				DepartOmniDirectFrm.FormClosed += Form_FormClosed;

				_tsmiDepartOmniDirect.Checked = true;
				_tsmiDepartRouts.Enabled = false;
				_tsmiDepartGuidance.Enabled = false;

				DepartOmniDirectFrm.Show(GlobalVars.Win32Window);
				DepartOmniDirectFrm.ComboBox001.Focus();
				GlobalVars.CurrCmd = 1;

				NativeMethods.HidePandaBox();
				DepartOmniDirectFrm = null;
			}
			catch (Exception ex)
			{
				NativeMethods.HidePandaBox();

				var tsmi = sender as ToolStripMenuItem;
				MessageBox.Show(ex.Message, tsmi.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
			}

		}

		private void DepartRouts_Click(object sender, EventArgs e)
		{
			if (GlobalVars.CurrCmd == 2)
				return;
			GlobalVars.CurrCmd = 0;

			try
			{
                if (!SlotSelector())
                    return;

                NativeMethods.ShowPandaBox(GlobalVars.gAranEnv.Win32Window.Handle.ToInt32());

				InitCommands();

				CDepartRouts DepartRoutsFrm = new CDepartRouts();

				DepartRoutsFrm.FormClosed += Form_FormClosed;

				_tsmiDepartOmniDirect.Enabled = false;
				_tsmiDepartRouts.Checked = true;
				_tsmiDepartGuidance.Enabled = false;

				DepartRoutsFrm.Show(GlobalVars.Win32Window);
				DepartRoutsFrm.ComboBox001.Focus();
				GlobalVars.CurrCmd = 2;

				NativeMethods.HidePandaBox();
				DepartRoutsFrm = null;
			}
			catch (Exception ex)
			{
				NativeMethods.HidePandaBox();

				var tsmi = sender as ToolStripMenuItem;
				MessageBox.Show(ex.Message, tsmi.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
			}
		}

		private void DepartGuidance_Click(object sender, EventArgs e)
		{
			if (GlobalVars.CurrCmd == 3)
				return;
			GlobalVars.CurrCmd = 0;

			try
			{
                if (!SlotSelector())
                    return;

                NativeMethods.ShowPandaBox(GlobalVars.gAranEnv.Win32Window.Handle.ToInt32());

				InitCommands();

				CDepartGuidanse DepartGuidanseFrm = new CDepartGuidanse();

				DepartGuidanseFrm.FormClosed += Form_FormClosed;

				_tsmiDepartOmniDirect.Enabled = false;
				_tsmiDepartRouts.Checked = true;
				_tsmiDepartGuidance.Enabled = false;

				DepartGuidanseFrm.Show(GlobalVars.Win32Window);
				DepartGuidanseFrm.ComboBox001.Focus();
				GlobalVars.CurrCmd = 3;
				NativeMethods.HidePandaBox();
				DepartGuidanseFrm = null;
			}
			catch (Exception ex)
			{
				NativeMethods.HidePandaBox();

				var tsmi = sender as ToolStripMenuItem;
				MessageBox.Show(ex.Message, tsmi.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
			}
		}

		void Form_FormClosed(Object sender, EventArgs e)
		{
			_tsmiDepartOmniDirect.Checked = false;
			_tsmiDepartRouts.Checked = false;
			_tsmiDepartGuidance.Checked = false;

			_tsmiDepartOmniDirect.Enabled = true;
			_tsmiDepartRouts.Enabled = true;
			_tsmiDepartGuidance.Enabled = true;

			GlobalVars.CurrCmd = 0;
			sender = null;
			GC.Collect();
		}
        #endregion

        public bool SlotSelector()
        {
            var slotSelector = true;

            var dbProvider = GlobalVars.gAranEnv.DbProvider as DbProvider;
            if (dbProvider != null && dbProvider.ProviderType == DbProviderType.TDB)
            {
                dynamic methodResult = new System.Dynamic.ExpandoObject();
                dbProvider.CallSpecialMethod("SelectSlot", methodResult);
                slotSelector = methodResult.Result;
            }

            if (!slotSelector)
            {
                MessageBox.Show("Please first select slot!", "Slot Selector", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                return slotSelector;
            }
            return slotSelector;
        }
	}
}
